libobs/obs-source.c

```diff
@@ -860,11 +861,16 @@ uint32_t obs_get_source_output_flags(const char *id)
 
 static void obs_source_deferred_update(obs_source_t *source)
 {
-	if (source->context.data && source->info.update)
-		source->info.update(source->context.data,
-				    source->context.settings);
-
-	source->defer_update = false;
+	if (source->context.data && source->info.update) {
+		long count;
+		do {
+			count = os_atomic_load_long(
+				&source->defer_update_count);
+			source->info.update(source->context.data,
+					    source->context.settings);
+		} while (!os_atomic_compare_swap_long(
+			&source->defer_update_count, count, 0));
```

Comment:
Alternatively if we are afraid of blocking for too long, we can update only once instead of the loop and let the next update be handled by the next video tick.

Reply:
If there are 10 deferred updates this loop does not look like it will apply each of the 10 updates one after another. It looks like it will instead apply the latest update 10 times.

Reply:
Oh i see its a swap 0 not a dec. Sorry.

Reply:
But yes i do think it should be deferred again if someone made a new change. Applying all the updates now would stall rendering.